src/supernova/mod.rs

```diff
@@ -0,0 +1,1142 @@
+      circuit_secondary.clone(),
+      num_augmented_circuit,
+    );
+
+    let test_circuit2 = SquareCircuit::new(OPCODE_1, rom.len());
+    let mut running_claim2 = RunningClaim::<
+      G1,
+      G2,
+      SquareCircuit<<G1 as Group>::Scalar>,
+      TrivialTestCircuit<<G2 as Group>::Scalar>,
+    >::new(
+      OPCODE_1,
+      test_circuit2,
+      circuit_secondary.clone(),
+      num_augmented_circuit,
+    );
+
+    // generate the commitkey based on max num of constraints and reused it for all other augmented circuit
+    let circuit_public_params = vec![&running_claim1.params, &running_claim2.params];
+    let (max_index_circuit, _) = circuit_public_params
+      .iter()
+      .enumerate()
+      .map(|(i, params)| -> (usize, usize) { (i, params.r1cs_shape_primary.num_cons) })
+      .max_by(|(_, circuit_size1), (_, circuit_size2)| circuit_size1.cmp(&circuit_size2))
+      .unwrap();
+
+    let ck_primary =
+      R1CS::<G1>::commitment_key(&circuit_public_params[max_index_circuit].r1cs_shape_primary);
+    let ck_secondary =
+      R1CS::<G2>::commitment_key(&circuit_public_params[max_index_circuit].r1cs_shape_secondary);
+
+    // set unified ck_primary, ck_secondary and update digest
+    running_claim1.params.ck_primary = Some(ck_primary.clone());
+    running_claim1.params.ck_secondary = Some(ck_secondary.clone());
+    running_claim1.params.digest =
+      compute_digest::<G1, PublicParams<G1, G2>>(&running_claim1.params);
+
+    running_claim2.params.ck_primary = Some(ck_primary.clone());
+    running_claim2.params.ck_secondary = Some(ck_secondary.clone());
+    running_claim2.params.digest =
+      compute_digest::<G1, PublicParams<G1, G2>>(&running_claim2.params);
+
+    let num_steps = rom.len();
+    let initial_program_counter = <G1 as Group>::Scalar::from(0);
+
+    // extend z0_primary/secondary with rom content
+    let mut z0_primary = vec![<G1 as Group>::Scalar::ONE]; // var, program_counter
+    z0_primary.extend(
+      rom
+        .iter()
+        .map(|opcode| <G1 as Group>::Scalar::from(*opcode as u64)),
+    );
+    let mut z0_secondary = vec![<G2 as Group>::Scalar::ONE]; // var, program_counter
+    z0_secondary.extend(
+      rom
+        .iter()
+        .map(|opcode| <G2 as Group>::Scalar::from(*opcode as u64)),
+    );
+
+    let mut recursive_snark_option: Option<RecursiveSNARK<G1, G2>> = None;
+
+    for _ in 0..num_steps {
+      let program_counter = recursive_snark_option
+        .as_ref()
+        .map(|recursive_snark| recursive_snark.program_counter)
+        .unwrap_or_else(|| initial_program_counter);
+      let augmented_circuit_index = rom[u32::from_le_bytes(
+        // convert program counter from field to usize (only took le 4 bytes)
+        program_counter.to_repr().as_ref().clone()[0..4]
+          .try_into()
+          .unwrap(),
+      ) as usize];
+
+      let mut recursive_snark = recursive_snark_option.unwrap_or_else(|| {
+        if augmented_circuit_index == OPCODE_0 {
```

Comment:
This demo have some code duplication on opcode multiplexing, due to slightly hard on handling heterogeneous type of augment circuit. I try few options to beautify this part and no satisfied version. 

If you have better suggestion kindly let me know, thanks